python/task_manager/background_task_manager.py

```diff
@@ -1,47 +1,142 @@
-        the ``result`` is the data returned by the task. 
-    
+        The ``uid`` parameter holds the unique id associated with the task,
+        the ``group`` is the group that the task is associated with and
+        the ``result`` is the data returned by the task.
+
     :signal task_failed(uid, group, message, traceback_str): Emitted when a task fails for some reason.
-        The ``uid`` parameter holds the unique id associated with the task, 
-        the ``group`` is the group that the task is associated with, 
+        The ``uid`` parameter holds the unique id associated with the task,
+        the ``group`` is the group that the task is associated with,
         the ``message`` is a short error message and the ``traceback_str``
         holds a full traceback.
-        
+
     :signal task_group_finished(group): Emitted when all tasks in a group have finished.
         The ``group`` is the group that has completed.
 
     """
 
+    class _ResultsPoller(QtCore.QObject):
+        """
+        Polls a queue which holds the results posted from the worker threads.
+        """
+
+        # Emitted when a task is completed.
+        task_completed = QtCore.Signal(object, object, object)
+        # Emitted when a task has failed.
+        task_failed = QtCore.Signal(object, object, object, object)
+
+        def __init__(self, parent=None):
+            """
+            Constructor.
+
+            :param parent:  The parent QObject for this thread
+            """
+            QtCore.QObject.__init__(self, parent)
+            # Results queue that will be polled
+            self._results = Queue.Queue()
+            # Create a timer with no interval. This means the timer will be invoked
+            # only when the event queue is empty.
+            self._timer = QtCore.QTimer()
+            self._timer.setInterval(100)
+            # This should be instantiated from the application's thread since we only support
+            # one instance for now.
+            assert(QtCore.QCoreApplication.instance().thread() == self._timer.thread())
+            # Since the event will be invoked as soon as the event queue is empty,
+            # we should process the queue one item at a time to not interfere with the
+            # ui responsivenes
+            self._timer.timeout.connect(self._flush_events)
+            self._bundle = sgtk.platform.current_bundle()
+
+        def start(self):
+            """
+            Start polling for tasks that ended.
+            """
+            self._timer.start()
+
+        def shut_down(self):
+            """
+            Stop polling for tasks that ended.
+            """
+            self._timer.stop()
+
+        def _flush_events(self):
+            """
+            Executes callbacks for each task result.
+            """
+            try:
+                # Get everything until the queue is empty.
+                while True:
+                    cb = self._results.get_nowait()
```

Comment:
get is blocking until something is available. We don't want that, since it would lock the event processing thread.